lib/Macros/Sources/SwiftMacros/PointerBoundsMacro.swift

```diff
@@ -0,0 +1,618 @@
+import SwiftDiagnostics
+import SwiftParser
+import SwiftSyntax
+import SwiftSyntaxBuilder
+import SwiftSyntaxMacros
+
+protocol ParamInfo: CustomStringConvertible {
+    var description: String { get }
+    var original: ExprSyntax { get }
+    var pointerIndex: Int { get }
+    var nonescaping: Bool { get set }
+
+    func getBoundsCheckedThunkBuilder(_ base: BoundsCheckedThunkBuilder, _ funcDecl: FunctionDeclSyntax,
+                                  _ variant: Variant
+                                 ) -> BoundsCheckedThunkBuilder
+}
+
+struct CountedBy: ParamInfo {
+    var pointerIndex: Int
+    var count: ExprSyntax
```

Comment:
How would you then represent count expressions with e.g. addition or multiplication?

Reply:
To get there, we'd have to "parse" the C/C++ expression and map it into something. We can represent that here with an expression in terms of Swift if we need to.

Reply:
Indeed. So far I've punted that issue, because the basic operations that cover most use cases have the same syntax in Swift. I imagine a macro syntax like this: `.mul(.param(2), .param(3))`. This has the upside of only being able to represent expressions we support, and downside of requiring us to add explicit support for every type of expression. It would also let us add an enum case representing the return value of a function, so malloc could be annotated with `@PointerBounds(.sizedByOrNull(pointer: .retVal, size: .param(1)))`.